content/articles/working-with-java-synchronization/index.md

```diff
@@ -0,0 +1,299 @@
+        t1.start();
+        t2.start();
+    }
+}
+```
+The above code will output the following
+
+```
+5
+10
+15
+20
+25
+100
+200
+300
+400
+500
+```
+
+The following is all that it entails:
+
+- The synchronized keyword is used to identify blocks that are part of a synchronized thread in Java. In Java, a synchronized block is one that is tied to a specific object. There can only be one thread operating in all synchronized blocks synchronized on the same object. When the synchronized block is exited, all subsequent threads trying to enter it are stalled until that thread quits.
+- Synchronized blocks utilize the synchronized keyword as a lock. When a method is marked as synchronized, the thread owns the monitor or lock object. In this case, you are blocked until the other thread releases the monitor.
+- Using a synchronized block enables you to fine-tune lock control by mutually excluding important section code.
+- The lock gets unlocked when the thread leaves the synchronized block.
+- A synchronized block may produce a `NullPointerException` if a parameter expression evaluates to null, while synchronized methods do not.
+- The lock is only held throughout the block scope of the synchronized block. 
+- A static method may lock an object inside the parenthesis of a synchronized block.
+
+### Implementing synchronization in Java
+To provide internal synchronization, Java's lock concept is employed. In Java, each object has its own lock. In this case, the lock idea will be brought into play anytime we utilize the synchronized keyword.
+
+First, a thread with the lock on the object must execute any synchronized methods on it. Post-lock, a thread may invoke any synchronized method on the object. Upon successful completion of the synchronized method, the thread is responsible for releasing the lock. No other threads may perform synchronized methods on the same object while a thread is executing one. However, any non-synchronized procedure may be executed concurrently by the remaining threads. Note that this idea of a lock may be applied on the object level rather than the method level.
+
+Let's look at an example program:
+
+```java
+class Implementation {
+
+	public synchronized void wish(String tag) {
+	    int x;
+		for ( x = 1; x <= 2; x++) {
+			System.out.println("Hello : ");
+			try {
+				Thread.sleep(100);
+			} catch (InterruptedException ignored) {
+
+			}
+			System.out.println(tag);
+		}
+	}
+
+}
+
+class OurThreadExample extends Thread {
+	Implementation b;
+	String tag;
+
+	public OurThreadExample(Implementation b, String tag) {
+		super();
+		this.b = b;
+		this.tag = tag;
+	}
+
+	public void run() {
+		b.wish(tag);
+	}
+}
+
+public class SynchImp {
+	public static void main(String[] args) {
+
+		Implementation b1 = new Implementation();
+		OurThreadExample mt1 = new OurThreadExample(b1, "SECTION");
+		OurThreadExample mt2 = new OurThreadExample(b1, "ENGINEERING");
+		mt1.start();
+		mt2.start();
+	}
+}

```

Comment:
This code also suffers from race conditions. Occasionally, it will output:

```
Hello : 
ENGINEERING
Hello : 
ENGINEERING
Hello : 
SECTION
Hello : 
SECTION
```

Reply:
Again, this can take quite a few executions to reproduce.